### PR TITLE
Add sp4 refactored

### PR DESCRIPTION
**Pull request by Tamarinen**
This adds support for Clas Ohlson SP4L-EU, as per @skrutt 's changes in #300 . I've just ported it to the even newer structure, and made sure it works as advertised.

### CODE REVIEW

**Comment by Tamarinen** [comment]
@mjg59 
I'm not sure why the "Merge pull request #1 from mjg59/master" commit was included. Will it be a problem?

File: broadlink/switch.py [hunk 1]
```diff
@@ -216,3 +216,97 @@ def get_energy(self) -> int:
...
+    """Controls a Broadlink SP4."""
+
+    def __init__(self, *args, **kwargs):
+        """Initialize the controller."""
+        device.__init__(self, *args, **kwargs)
+        self.type = "SP4"
+
+    def set_power(self, state):
+        """Set the power state of the device"""
+        return self.set_state(pwr=int(bool(state)))
```
Comment (felipediel) [inline broadlink/switch.py:231]: We are duck typing now. Please annotate the expected input and remove unnecessary conversions.
## Example
### Before
```python3
    def set_power(self, state):
        """Set the power state of the device"""
        return self.set_state(pwr=int(bool(state)))
```
### After
```python3
    def set_power(self, state: bool) -> bool:
        """Set the power state of the device"""
        return self.set_state(pwr=state)
```

File: broadlink/switch.py [hunk 2]
```diff
@@ -216,3 +216,97 @@ def get_energy(self) -> int:
...
+        return self.set_state(ntlight=int(bool(state)))
+
+    def set_state(
+        self,
+        pwr=None,
+        ntlight=None,
+        indicator=None,
+        ntlbrightness=None,
+        maxworktime=None,
+    ):
```
Comment (felipediel) [inline broadlink/switch.py:244]: Please annotate input and return value.
## Example
```python3
    def set_state(
        self,
        pwr: bool = None,
        ntlight: bool = None,
        indicator: bool = None,
        ntlbrightness: int = None,
        maxworktime: int = None,
    ) -> dict:
```

File: broadlink/switch.py [hunk 3]
```diff
@@ -216,3 +216,97 @@ def get_energy(self) -> int:
...
+        """Set state of device"""
+        data = {}
+        if pwr is not None:
+            data["pwr"] = int(bool(pwr))
```
Comment (felipediel) [inline broadlink/switch.py:248]: Remove unnecessary conversions. The client should provide an input according to the annotations.

**Review by felipediel — CHANGES_REQUESTED** [review]
Thank you very much for not letting this update die. Some considerations that will help us to maintain and document the code further...

File: broadlink/switch.py [hunk 4]
```diff
@@ -260,19 +260,17 @@ def set_state(
         response = self.send_packet(0x6A, packet)
         return self._decode(response)
 
-    def check_power(self):
+    def check_power(self) -> bool:
         """Return the power state of the device."""
         state = self.get_state()
-        if state:
-            return state["pwr"]
+        return state["pwr"] if state else None
```
Comment (felipediel) [inline broadlink/switch.py:266]: Why not just `return state["pwr"]`?

Comment (Tamarinen) [reply broadlink/switch.py:266]: The original code checked if state is truthy. I imagine that was a safeguard against `get_state()` not returning the expected dict, in which case there wouldn't be any `pwr`to get. `get_state()`, in turn, ultimately depends on `json.loads()`, which on closer inspection shouldn't be able to fail in a way that produce a falsy result, but rather a `JSONDecodeError`. So yes, those lines should be perfectly ok without the safeguard. I'll make them so.

Comment (felipediel) [reply broadlink/switch.py]: I think these functions should return None. If something goes wrong the device will return an error in the response and we raise [here](https://github.com/mjg59/python-broadlink/blob/d8b853d96c9bc64a253e41657789359f6bb165bd/broadlink/switch.py#L305). So there is no reason to return a state we already know.

File: broadlink/switch.py [hunk 5]
```diff
@@ -221,35 +221,35 @@ def get_energy(self) -> int:
...
     """Controls a Broadlink SP4."""
 
-    def __init__(self, *args, **kwargs):
+    def __init__(self, *args, **kwargs) -> None:
         """Initialize the controller."""
         device.__init__(self, *args, **kwargs)
         self.type = "SP4"
 
-    def set_power(self, state):
+    def set_power(self, state: bool) -> bool:
```
Comment (felipediel) [inline broadlink/switch.py:229]: These functions should return None. If something goes wrong the device will return an error in the response and we raise [here](https://github.com/mjg59/python-broadlink/blob/d8b853d96c9bc64a253e41657789359f6bb165bd/broadlink/switch.py#L305). So there is no reason to return a state we already know. Follow the SP2 template.

File: broadlink/switch.py [hunk 6]
```diff
@@ -221,35 +221,35 @@ def get_energy(self) -> int:
...
         """Set the power state of the device"""
-        return self.set_state(pwr=int(bool(state)))
+        return self.set_state(pwr=state)
 
-    def set_nightlight(self, state):
+    def set_nightlight(self, state: bool) -> bool:
```
Comment (felipediel) [inline broadlink/switch.py:233]: Same here.

File: broadlink/switch.py [hunk 7]
```diff
@@ -221,35 +221,35 @@ def get_energy(self) -> int:
...
-        indicator=None,
-        ntlbrightness=None,
-        maxworktime=None,
-    ):
+        pwr: bool = None,
+        ntlight: bool = None,
+        indicator: bool = None,
+        ntlbrightness: int = None,
+        maxworktime: int = None,
+    ) -> dict:
```
Comment (felipediel) [inline broadlink/switch.py:244]: Here is fine, because we get information we didn't send.

File: broadlink/switch.py [hunk 8]
```diff
@@ -216,3 +216,95 @@ def get_energy(self) -> int:
...
+
+    def set_state(
+        self,
+        pwr: bool = None,
+        ntlight: bool = None,
+        indicator: bool = None,
+        ntlbrightness: int = None,
+        maxworktime: int = None,
+    ) -> dict:
+        """Set state of device"""
```
Comment (felipediel) [inline broadlink/switch.py:245]: Add "."

**Comment by Tamarinen** [comment]
Thanks for excellent help and feedback!

File: broadlink/switch.py [hunk 9]
```diff
@@ -216,3 +216,95 @@ def get_energy(self) -> int:
...
+class sp4(device):
+    """Controls a Broadlink SP4."""
+
+    def __init__(self, *args, **kwargs) -> None:
+        """Initialize the controller."""
+        device.__init__(self, *args, **kwargs)
+        self.type = "SP4"
+
+    def set_power(self, state: bool) -> None:
+        """Set the power state of the device"""
```
Comment (felipediel) [inline broadlink/switch.py:230]: Add "."

File: broadlink/switch.py [hunk 10]
```diff
@@ -216,3 +216,95 @@ def get_energy(self) -> int:
...
+        self.set_state(pwr=state)
+
+    def set_nightlight(self, state: bool) -> None:
+        """Set the night light state of the device"""
```
Comment (felipediel) [inline broadlink/switch.py:234]: Add .

File: broadlink/switch.py [hunk 11]
```diff
@@ -216,3 +216,95 @@ def get_energy(self) -> int:
...
+        state = self.get_state()
+        return state["pwr"]
+
+    def check_nightlight(self) -> bool:
+        """Return the night light state of the device."""
+        state = self.get_state()
+        return state["ntlight"]
+
+    def get_state(self) -> dict:
+        """Get full state of device"""
```
Comment (felipediel) [inline broadlink/switch.py:274]: Add .

**Comment by felipediel** [comment]
Beautiful. Is this code working?

**Comment by Tamarinen** [comment]
Works for me. :) I've used it, in several iterations, to switch the charging of my computer on and off in order to save at least some of the NON-REMOVABLE!! battery's capacity instead of just having it at 100% permanently. I'm altering the nightlight brightness, too, to indicate that the computer and network is still working.

It might be worth asking @skrutt if he still has his plug and would like to test-drive it.

**Comment by felipediel** [comment]
Thank you very much! I will take this to Home Assistant in the future. If someone's house catches on fire, I'll send your phone.

@mjg59 I made your life a little easier here. You can pass this update in front of mine. I'm still polishing and splitting those things. If necessary, I'll do a rebase later, but I don't think it will be.

**Comment by mjg59** [comment]
Awesome, thank you!

**Comment by Tamarinen** [comment]
@felipediel @mjg59 

Thanks, this will make keeping up to date easier for me!

The full status message hints at there being SP4 devices with energy monitoring and USB connectors, but as mine doesn't have or do, I didn't bother with code I couldn't really test. With the SP4 device in place, the barrier to add those details by people who has devices with those capabilities is considerably lower.

` {'pwr': 0, 'ntlight': 1, 'indicator': 1, 'usbpwr': 0, 'maxworktime': 0, 'usbmaxworktime': 0, 'ntlbrightness': 59, 'current': 0, 'volt': -1, 'power': 0, 'totalconsum': -1, 'overload': -1}`

**Comment by felipediel** [comment]
Ok, thank you!

**Pull request merged**